dirty_cat/minhash_encoder.py

```diff
@@ -172,56 +177,47 @@ def transform(self, X):
-                    else:
-                        X_out[i, k * self.n_components:counter] = self.hash_dict[x]
-                counter += self.n_components
-        elif self.hashing == 'murmur':
-            X_out = np.zeros((len(X[:]), self.n_components * X.shape[1]))
-            counter = self.n_components
-            for k in range(X.shape[1]):
-                X_in = X[:, k].reshape(-1)
-                for i, x in enumerate(X_in):
-                    if isinstance(x, float):
-                        is_nan_idx = True
-                    elif x not in self.hash_dict:
-                        X_out[i, k * self.n_components:counter] = self.hash_dict[x] = self.minhash(
+        if not (X == X).all():  # contains at least one missing value
+            if self.handle_missing == 'error':
+                raise ValueError("Found missing values in input data; set "
+                                 "handle_missing='zero_impute' to encode with missing values")
+            elif self.handle_missing == 'zero_impute':
+                X[~(X == X)] = "NAN"  # this will be replaced by a vector of zeroes by compute_hash
+
+        def compute_hash(x):
+            # Function called by joblib to compute the hash
+            if x not in self.hash_dict:
+                if x == "NAN":  # true if x is a missing value
+                    self.hash_dict[x] = np.zeros(self.n_components)
+                else:
+                    if self.hashing == "fast":
+                        self.hash_dict[x] = self.get_fast_hash(x)
+                    elif self.hashing == "murmur":
+                        self.hash_dict[x] = self.minhash(
                             x,
                             n_components=self.n_components,
                             ngram_range=self.ngram_range
                         )
                     else:
-                        X_out[i, k * self.n_components:counter] = self.hash_dict[x]
-                counter += self.n_components
-        else:
-            raise ValueError("hashing function must be 'fast' or"
-                             "'murmur', got '{}'"
-                             "".format(self.hashing))
-
-        if self.handle_missing == 'error' and is_nan_idx:
-            msg = ("Found missing values in input data; set "
-                   "handle_missing='zero_impute' to encode with missing values")
-            raise ValueError(msg)
-        return X_out
+                        raise ValueError("hashing function must be 'fast' or"
+                                         "'murmur', got '{}'"
+                                         "".format(self.hashing))
+            return self.hash_dict[x]
+
+        # Compute the hashes for unique values
+        unique_x, indices_x = np.unique(X, return_inverse=True)
+        unique_x_trans = Parallel(n_jobs=self.n_jobs)(delayed(compute_hash)(x) for x in unique_x)
```

Comment:
I think that we should apply things on batches rather than on single rows. I fear that single rows will lead to overhead.

Maybe we should to to use 
https://scikit-learn.org/stable/modules/generated/sklearn.utils.gen_even_slices.html
as in
https://github.com/scikit-learn/scikit-learn/blob/703bee65e2122ad273cf0b42460c5c28ca638af8/sklearn/decomposition/_lda.py#L460

We should benchmark this to see what the fastest / best option. Talk to @jjerphan, he has good experience in benchmarking parallel computing.

Reply:
Benchmarking really quickly on my Mac M1 with 8 cores, the rows version is about twice faster than the batched version, but we should do a more serious benchmark.